controller/metrics/prometheus.py

```diff
@@ -0,0 +1,227 @@
+from typing import Any, Dict, List, Union
+from prometheus_client import Counter, Histogram, Summary, Gauge
+from dateutil import parser
+import re
+
+from controller.server_status_enum import ServerStatusEnum
+from controller.metrics.events import MetricEventHandler, MetricEvent
+from controller.metrics.utils import resource_request_from_manifest
+
+
+class PrometheusMetric():
+    _types = {
+        "counter": {
+            "usage_methods": ["inc"],
+            "type": Counter,
+        },
+        "gauge": {
+            "usage_methods": ["inc", "dec", "set"],
+            "type": Gauge,
+        },
+        "histogram": {
+            "usage_methods": ["observe"],
+            "type": Histogram,
+        },
+        "summary": {
+            "usage_methods": ["observe"],
+            "type": Summary,
+        },
+    }
+
+    def __init__(
+        self,
+        metric_type: str,
+        name: str,
+        documentation: str,
+        labelnames: List[str] = [],
+        *args,
+        **kwargs
+    ):
+        if metric_type not in self._types:
+            raise ValueError(
+                f"Metric type {metric_type} not found in allowed types: {self._types.keys()}."
+            )
+        self.metric_type = metric_type
+        self.usage_methods = self._types[metric_type]["usage_methods"]
+        sanitized_label_names = [self.sanitize_label_name(label) for label in labelnames]
+        self._metric = self._types[metric_type]["type"](
+            name, documentation, sanitized_label_names, *args, **kwargs
+        )
+
+    @staticmethod
+    def sanitize_label_name(val: str) -> str:
+        invalid_first_letter = re.compile(r"^[^a-zA-Z_]")
+        invalid_all_letters = re.compile(r"[^a-zA-Z0-9_]")
+        val = re.sub(invalid_first_letter, "_", val, count=1)
+        val = re.sub(invalid_all_letters, "_", val)
+        return val
+
+    @classmethod
+    def _sanitize_labels(cls, labels: Dict[str, str]) -> Dict[str, str]:
+        sanitized_labels = {}
+        for name, val in labels.items():
+            sanitized_labels[cls.sanitize_label_name(name)] = val
+        return sanitized_labels
```

Comment:
I added that because they do did not really need `self` of any of the real state of the object. But with some of the other proposed changes now I do need access to the state and there is no more classmethod or staticmethod